File: server/src/scripts/check_tables.ts
import dotenv from "dotenv";
import { resolve, dirname } from "path";
import { fileURLToPath } from "url";
import { pool } from "../lib/postgres.js";

const __filename = fileURLToPath(import.meta.url);
const __dirname = dirname(__filename);
dotenv.config({ path: resolve(__dirname, "../../../.env") });

async function verifyAllTables() {
  console.log("--- STARTING DATABASE SCHEMAS VERIFICATION ---");
  try {
    // 1. Get current database & user details
    const connInfo = await pool.query("SELECT current_database() as db, current_user as usr, version() as ver");
    console.log(`Connected to Database: "${connInfo.rows[0].db}" as User: "${connInfo.rows[0].usr}"`);
    console.log(`PostgreSQL Version: ${connInfo.rows[0].ver.split(",")[0]}`);
    console.log("----------------------------------------------");

    // 2. Fetch list of tables in public schema
    const tablesRes = await pool.query(`
      SELECT table_name 
      FROM information_schema.tables 
      WHERE table_schema = 'public' AND table_type = 'BASE TABLE'
      ORDER BY table_name;
    `);

    console.log(`Found ${tablesRes.rows.length} base tables in 'public' schema:`);

    for (const row of tablesRes.rows) {
      const tableName = row.table_name;
      
      // Get column count
      const colRes = await pool.query(`
        SELECT count(*)::int as count 
        FROM information_schema.columns 
        WHERE table_schema = 'public' AND table_name = $1
      `, [tableName]);

      // Get row count estimate
      const countRes = await pool.query(`SELECT count(*)::int as count FROM "${tableName}"`);

      // Get indexes for table
      const idxRes = await pool.query(`
        SELECT indexname, indexdef 
        FROM pg_indexes 
        WHERE schemaname = 'public' AND tablename = $1
      `, [tableName]);

      console.log(`\n🔹 Table: "${tableName}"`);
      console.log(`   - Columns: ${colRes.rows[0].count}`);
      console.log(`   - Rows:    ${countRes.rows[0].count}`);
      console.log(`   - Indexes: ${idxRes.rows.length}`);
      for (const idx of idxRes.rows) {
        console.log(`     * ${idx.indexname}: ${idx.indexdef}`);
      }
    }

    console.log("\n----------------------------------------------");
    console.log("✅ DATABASE SCHEMA AUDIT COMPLETE");

  } catch (err: any) {
    console.error("❌ Schema verification failed:", err.message);
  } finally {
    await pool.end();
  }
}

verifyAllTables();

File: server_py/scripts/check_tables.py
import os
import sys
import psycopg2

# Setup import path
base_dir = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
sys.path.append(base_dir)

from app.config import DATABASE_URL

def verify_all_tables() -> None:
    print("--- STARTING DATABASE SCHEMAS VERIFICATION ---")
    if not DATABASE_URL:
        print("DATABASE_URL is not set!")
        sys.exit(1)
        
    try:
        conn = psycopg2.connect(DATABASE_URL)
        cur = conn.cursor()
        
        # 1. Get database details
        cur.execute("SELECT current_database(), current_user, version();")
        db_name, db_user, db_version = cur.fetchone()
        print(f"Connected to Database: \"{db_name}\" as User: \"{db_user}\"")
        print(f"PostgreSQL Version: {db_version.split(',')[0]}")
        print("----------------------------------------------")
        
        # 2. Fetch list of tables
        cur.execute("""
            SELECT table_name 
            FROM information_schema.tables 
            WHERE table_schema = 'public' AND table_type = 'BASE TABLE'
            ORDER BY table_name;
        """)
        tables = [row[0] for row in cur.fetchall()]
        print(f"Found {len(tables)} base tables in 'public' schema:")

        for table in tables:
            # Columns count
            cur.execute("""
                SELECT count(*) 
                FROM information_schema.columns 
                WHERE table_schema = 'public' AND table_name = %s
            """, (table,))
            cols_count = cur.fetchone()[0]

            # Rows count
            cur.execute(f'SELECT count(*) FROM "{table}"')
            rows_count = cur.fetchone()[0]

            # Indexes
            cur.execute("""
                SELECT indexname, indexdef 
                FROM pg_indexes 
                WHERE schemaname = 'public' AND tablename = %s
            """, (table,))
            indexes = cur.fetchall()

            print(f'\n\ud83d\udd39 Table: "{table}"')
            print(f"   - Columns: {cols_count}")
            print(f"   - Rows:    {rows_count}")
            print(f"   - Indexes: {len(indexes)}")
            for idx_name, idx_def in indexes:
                print(f"     * {idx_name}: {idx_def}")

        print("\n----------------------------------------------")
        print("\u2705 DATABASE SCHEMA AUDIT COMPLETE")
        cur.close()
        conn.close()

    except Exception as err:
        print(f"\u274c Schema verification failed: {err}")

if __name__ == "__main__":
    verify_all_tables()
